homeworks/08-functional-reactive-programming/src/index.ts

```diff
@@ -0,0 +1,107 @@
+import {BehaviorSubject, fromEvent, timer} from "rxjs";
+import {map, scan, startWith, switchMap, tap} from "rxjs/operators";
+
+import "./index.scss";
+import {
+    addRandomRoach,
+    addRandomWindows, getItemHeight, getItemWidth,
+    getRoachPosition, ICell, IGameField,
+    imagesLoaded,
+    initializeGameField,
+    IRoachPosition, removeRoach
+} from "./utils";
+import {Add, IAction, Null, stateReducer} from "./state/state";
+
+const $root = document.querySelector<HTMLDivElement>("#root");
+const $canvas = document.querySelector<HTMLCanvasElement>("#canvas");
+const $btnNewHouse = document.querySelector<HTMLButtonElement>("#btnNewHouse");
+const $score = document.querySelector<HTMLSpanElement>("#score");
+const context = $canvas?.getContext("2d");
+const $imageWall = new Image();
+$imageWall.src = require("./svg/brickwall.svg").default;
+const $imageWindow = new Image();
+$imageWindow.src = require("./svg/window.svg").default;
+const $imageRoach = new Image();
+$imageRoach.src = require("./svg/roach.svg").default;
+
+imagesLoaded($imageWindow, $imageWall, $imageRoach).then(() => {
+    if ($root && $canvas && context && $btnNewHouse && $score) {
+        $canvas.style.width = "500px";
+        $canvas.style.height = "500px";
+        const canvasRect = $canvas.getBoundingClientRect();
+        const canvasScale = window.devicePixelRatio;
+        $canvas.width  = canvasRect.width  * canvasScale;
+        $canvas.height = canvasRect.height * canvasScale;
+        context?.scale(canvasScale, canvasScale);
+        const {width, height} = $canvas.getBoundingClientRect();
+
+        const actions$ = new BehaviorSubject<IAction>(Add());
+        const state$ = actions$
+            .pipe(
+                startWith(0),
+                scan(stateReducer)
+            )
+            .subscribe((value) => {
+                $score.textContent = value.toString();
+            })
+
+        const subject = new BehaviorSubject<IRoachPosition>(null);
+        const gameField$ = fromEvent($btnNewHouse, "click")
+            .pipe(
+                startWith(""),
+                map(initializeGameField(10, 8)),
+                map(initializeGameField(10, 8)),
+                map(addRandomWindows(10)),
```

Comment:
If you Use RxJS operators you should keep data immutable. Each operator should return new data depending on input data. You shouldn't mutate them.